tests/codegen/atomicptr.rs

```diff
@@ -0,0 +1,38 @@
+// LLVM does not support some atomic RMW operations on pointers, so inside codegen we lower those
+// to integer atomics, surrounded by casts to and from integer type.
+// This test ensures that we do the round-trip correctly for AtomicPtr::fetch_byte_add, and also
+// ensures that we do not have such a round-trip for AtomicPtr::swap, because LLVM supports pointer
+// arguments to `atomicrmw xchg`.
+
+//@ compile-flags: -O -Cno-prepopulate-passes
+#![crate_type = "lib"]
+
+#![feature(strict_provenance)]
+#![feature(strict_provenance_atomic_ptr)]
+
+use std::sync::atomic::AtomicPtr;
+use std::sync::atomic::Ordering::Relaxed;
+use std::ptr::without_provenance_mut;
+
+// Portability hack so that we can say [[USIZE]] instead of i64/i32/i16 for usize.
+// CHECK: @helper([[USIZE:i[0-9]+]] noundef %_1)
+#[no_mangle]
+pub fn helper(_: usize) {}
+
+// CHECK-LABEL: @atomicptr_fetch_byte_add
+#[no_mangle]
+pub fn atomicptr_fetch_byte_add(a: &AtomicPtr<u8>, v: usize) -> *mut u8 {
+    // CHECK: %[[INTPTR:.*]] = ptrtoint ptr %{{.*}} to [[USIZE]]
+    // CHECK-NEXT: %[[RET:.*]] = atomicrmw add ptr %{{.*}}, [[USIZE]] %[[INTPTR]]
+    // CHECK-NEXT: inttoptr [[USIZE]] %[[RET]] to ptr
```

Comment:
I think this would be clearer (and not too fragile) if you match the whole thing end-to-end, i.e.
```suggestion
    // CHECK: %[[V_PTR:.*]] = inttoptr [[USIZE]] %v to ptr
    // CHECK: %[[V_INT:.*]] = ptrtoint ptr %[[V_PTR]] to [[USIZE]]
    // CHECK: %[[RET_INT:.*]] = atomicrmw add ptr %a, [[USIZE]] %[[V_INT]]
    // CHECK: %[[RET_PTR:.*]] = inttoptr [[USIZE]] %[[RET]] to ptr
    // CHECK: ret ptr %[[RET_PTR]]
```
(Could also match the argument names from the signature instead of hardcoding %a and %v, but I think it's fine, it's pretty unlikely that LLVM will start changing/dropping them in the future.)

Reply:
This then leads to the question--why do we roundtrip %v through inttoptr/ptrtoint?

I suppose this is because in MIR the atomic op takes two pointer arguments, but does it really make sense to add two pointers?

Reply:
Correct, adding two pointers doesn't make sense. The problem is that we're implementing an operation which takes a pointer operand and an integer operand. There's no atomic ptradd.

In terms of provenance, this is cursed if there are two pointer operands because which provenance is written back through the pointer? It's also cursed if we do this with integers, because then haven't we written an integer over a pointer, wiping the provenance from the memory? (I'm sort of hand-waving the kind of provenance semantics in Miri over LLVM IR which is itself a bit dubious)

Reply:
Right, that makes sense. I should've been more explicit--what I mean is, the relevant part of the [MIR](https://godbolt.org/z/dsnx468Kr) for your example currently looks like:
```rust
fn atomicptr_fetch_byte_add(_1: &AtomicPtr<u8>, _2: usize) -> *mut u8 {
  ...
  _4 = move _7 as *mut *mut u8 (PtrToPtr);
  _6 = _2 as *mut u8 (Transmute);
  _3 = atomic_xadd_relaxed::<*mut u8>(move _4, move _6) -> [return: bb1, unwind unreachable];
```
i.e. the intrinsic we're implementing here does take two pointer operands. Isn't that wrong?

Reply:
In what way would an intrinsic based on a pointer to an integer and and an integer operand be better? Or are you suggesting that we add an intrinsic `fn atomic_ptradd_relaxed(*mut *mut T, usize)`?

Reply:
> Or are you suggesting that we add an intrinsic fn atomic_ptradd_relaxed(*mut *mut T, usize)?

Yes.

Given that the only sensible operation here is offsetting a pointer by an integer, surely the intrinsic should match that?

(I guess this isn't so much a problem because we can just define `atomic_xadd_relaxed::<*mut u8>` to always use the provenance of the first argument, and I assume this is what Miri implements. Having to convert the offset to/from a pointer "just" makes the codegen more complex.)

Reply:
> always use the provenance of the first argument, and I assume this is what Miri implements.

Yes [indeed](https://github.com/rust-lang/miri/blob/e6bc750cb39a68805748dc205aae929befbd97ca/src/operator.rs#L64-L68).

But ideally we'd convince LLVM to provide an intrinsic that takes `(*mut *mut T, usize)`; that is the only way to avoid pointer-integer transmutation.

Reply:
Right, we need the LLVM change regardless, I don't mean to imply that this would solve the provenance issue, it would just remove some other unnecessary weirdness.

It doesn't need to be done as part of this PR though, I can do it separately.

Reply:
Agree with @erikdesjardins that the Rust intrinsic should accept usize rather than pointer for the second arg.